expr/func_aspercent.go

```diff
@@ -0,0 +1,241 @@
+	totalSeries GraphiteFunc
+	nodes       []expr
+}
+
+func NewAsPercent() GraphiteFunc {
+	return &FuncAsPercent{totalFloat: math.NaN()}
+}
+
+func (s *FuncAsPercent) Signature() ([]Arg, []Arg) {
+	return []Arg{
+		ArgSeriesList{val: &s.in},
+		ArgIn{
+			key: "total",
+			opt: true,
+			args: []Arg{
+				ArgFloat{val: &s.totalFloat},
+				ArgSeriesList{val: &s.totalSeries},
+			},
+		},
+		ArgStringsOrInts{val: &s.nodes, opt: true, key: "nodes"},
+	}, []Arg{ArgSeriesList{}}
+}
+
+func (s *FuncAsPercent) Context(context Context) Context {
+	return context
+}
+
+func (s *FuncAsPercent) Exec(cache map[Req][]models.Series) ([]models.Series, error) {
+	series, err := s.in.Exec(cache)
+	if err != nil {
+		return nil, err
+	}
+
+	var outSeries []models.Series
+	var totals []models.Series
+	if s.totalSeries != nil {
+		totals, err = s.totalSeries.Exec(cache)
+		if err != nil {
+			return nil, err
+		}
+		if len(totals) == 0 {
+			totals = nil
+		}
+	}
+
+	if s.nodes != nil {
+		// Set of keys
+		keys := make(map[string]struct{})
+		// Series grouped by key
+		metaSeries := groupSeriesByKey(series, s.nodes, &keys)
+		// The totals series for each key
+		var totalSeries map[string]models.Series
+
+		// calculate the sum
+		if math.IsNaN(s.totalFloat) && totals == nil {
+			totalSeries = getTotalSeries(metaSeries)
+			// calculate sum of totals series
+		} else if totals != nil {
+			totalSeriesLists := groupSeriesByKey(totals, s.nodes, &keys)
+			totalSeries = getTotalSeries(totalSeriesLists)
+		} else {
+			return nil, errors.New("total must be None or a seriesList")
+		}
+
+		for key := range keys {
+			// No series for total series
+			if _, ok := metaSeries[key]; !ok {
+				serie2 := totalSeries[key]
+				serie2.QueryPatt = fmt.Sprintf("asPercent(MISSING,%s)", serie2.QueryPatt)
+				serie2.Target = fmt.Sprintf("asPercent(MISSING,%s)", serie2.Target)
+				serie2.Tags = map[string]string{"name": serie2.Target}
+				for i := range serie2.Datapoints {
+					serie2.Datapoints[i].Val = math.NaN()
+				}
+				outSeries = append(outSeries, serie2)
+				cache[Req{}] = append(cache[Req{}], serie2)
+				continue
+			}
+
+			for _, serie1 := range metaSeries[key] {
+				// no total
+				if _, ok := totalSeries[key]; !ok {
+					serie1.QueryPatt = fmt.Sprintf("asPercent(%s,MISSING)", serie1.QueryPatt)
+					serie1.Target = fmt.Sprintf("asPercent(%s,MISSING)", serie1.Target)
+					serie1.Tags = map[string]string{"name": serie1.Target}
+					for i := range serie1.Datapoints {
+						serie1.Datapoints[i].Val = math.NaN()
+					}
+				} else {
+					serie2 := totalSeries[key]
+					serie1.QueryPatt = fmt.Sprintf("asPercent(%s,%s)", serie1.QueryPatt, serie2.QueryPatt)
+					serie1.Target = fmt.Sprintf("asPercent(%s,%s)", serie1.Target, serie2.Target)
+					serie1.Tags = map[string]string{"name": serie1.Target}
+					for i := range serie1.Datapoints {
+						serie1.Datapoints[i].Val = computeAsPercent(serie1.Datapoints[i].Val, serie2.Datapoints[i].Val)
```

Comment:
I think this violates https://github.com/grafana/metrictank/blob/master/expr/NOTES#L35